dask_glm/utils.py

```diff
@@ -140,6 +140,10 @@ def sum(A):
 def add_intercept(X):
     return np.concatenate([X, np.ones((X.shape[0], 1))], axis=1)
 
+@dispatch(cupy.ndarray)
```

Comment:
What about this cupy dispatch, do you think we should move it elsewhere?

Reply:
Yes, the dask-glm library needs to remain usable even if users don't have cupy installed.  Most users won't have a GPU.  

Perhaps we can generalized the NumPy solution above to look more like this, and then replace the dispatch type of that function with `object` rather `np.ndarray` so that it covers both.  This will also get triggered then if this function is called on non-array objects, but that would be an error anyway.

Reply:
This might be a common theme for other functions in this file as well.  We might try to find general implementations for the other functions as well.  

Looking at this code my *guess* is that the other `object` dispatches here are trying to support scipy.sparse matrices like 

```python
import scipy.sparse
x = scipy.sparse.csr_matrix([[1, 2], [0, 1]])
```

In some cases these extra dispatches are unnecessary, as in this case:

```python
@dispatch(object)
def log1p(A):
    return A.log1p()


@dispatch(np.ndarray)
def log1p(A):
    return np.log1p(A)

# where really the numpy implementation would have worked for either
In [11]: np.log1p(x)
Out[11]:
<2x2 sparse matrix of type '<class 'numpy.float64'>'
	with 3 stored elements in Compressed Sparse Row format>
```

In these cases we can probably replace the dispatched function with just plain `np.log1p`.  The `absolute` function can be replaced with just `abs` (Python dispatches on this itself).  I suspect that neither sign nor exp will work well for sparse arrays, so maybe we just drop support for them and use the numpy versions everywhere as well.

Reply:
I'm not sure I understand, are you suggesting we drop all these dispatches that just call `np.some_function()` and call `np.some_function()` directly?

Reply:
Yes, I think that in many (but not all) cases that will work just fine.  We don't need the dispatch between numpy and dask array.  The `object` dispatch is also often unnecessary.

```python
In [1]: import numpy as np

In [2]: import dask.array as da

In [3]: import scipy.sparse

In [4]: x = scipy.sparse.csr_matrix([[1, 2], [0, 1]])

In [5]: d = da.from_array(x, chunks='auto')

In [6]: np.sign(d)
Out[6]: dask.array<sign, shape=(2, 2), dtype=int64, chunksize=(2, 2)>

In [7]: np.log1p(d)
Out[7]: dask.array<log1p, shape=(2, 2), dtype=float64, chunksize=(2, 2)>

In [8]: abs(d)
Out[8]: dask.array<abs, shape=(2, 2), dtype=int64, chunksize=(2, 2)>

In [9]: np.dot(d, d)  # this seems not to work, it seems to return a computed array
Out[9]:
<2x2 sparse matrix of type '<class 'numpy.int64'>'
	with 3 stored elements in Compressed Sparse Row format>
```

Reply:
```diff
-@dispatch(object)
-def log1p(A):
-    return A.log1p()


-@dispatch(np.ndarray)
-def log1p(A):
-    return np.log1p(A)


-@dispatch(da.Array)
-def log1p(A):
-    return da.log1p(A)

+from numpy import log1p
```

Or, probably better, just use `np.log1p` in the code rather than `utils.log1p`

Reply:
I see, I'll check the ones we can get rid of against the current test set we have in-place.

Reply:
I just attempted removing all `dot()` dispatchers and running tests, everything passed. I'm only not sure whether we should keep some specializations for Dask arrays, like this one:

```python
@dispatch(da.Array, np.ndarray)
def dot(A, B):
    B = da.from_array(B, chunks=B.shape)
    return da.dot(A, B)
```

Since this is using Dask chunks, performance may be negatively impacted on cases that do take advantage of this. Any thoughts?

Reply:
We want to keep the dot dispatch

```python
In [9]: np.dot(d, d)  # this seems not to work, it seems to return a computed array
Out[9]:
<2x2 sparse matrix of type '<class 'numpy.int64'>'
	with 3 stored elements in Compressed Sparse Row format>
```

This failed because it returned a computed value rather than a lazy dask array value.  If this succeeded we would have seen an output like 

    <dask.array<dot, shape=(2, 2), dtype=int64, chunksize=(2, 2)>

Reply:
Alright, I'll leave the `dot()` dispatchers as they are now.